### PR TITLE
DDF-5156 Upgrades styled components to v4 (catalog-ui-search)

### PR DESCRIPTION
**Pull request by andrewkfiedler**
- Upgrades to the latest styled components in catalog-ui-search.  Ran into issues upgrading imperios instance, which we should revisit.  It brings it's own version of styled to the page, so we should look at fixing it's dependencies to avoid this.
 - Updates node / yarn to the latest versions.
 - Updates lerna to the latest version.
 - Had to add a specific version to catalog-ui-search-e2e's ace because it was pulling ace from npm for some reason.
 - See https://www.styled-components.com/docs/faqs#what-do-i-need-to-do-to-migrate-to-v4
 - See https://spectrum.chat/styled-components/general/typescript-support-for-refs~76750641-e0de-4fdb-82b8-f09aa5ca42f1 where I'm looking for how to address some of the changes I had to make.

### CODE REVIEW

**Comment by andrewkfiedler** [comment]
build now

File: ui/package.json [hunk 1]
```diff
@@ -29,22 +29,28 @@
...
+    ],
+    "nohoist": [
+      "**/styled-components"
+    ]
+  },
   "resolutions": {
     "admin-ui/@connexta/atlas/backbone.marionette": "1.8.8",
-    "@connexta/ace": "git://github.com/connexta/ace.git#d48cdcd457289b66a310f2ab7ed4564ce5d8db93"
-  }
+    "@connexta/ace": "git://github.com/connexta/ace.git#235d5b3885e1525e2f42bfbd64cd9715d85f498c"
```
Comment (Bdthomson) [inline ui/package.json]: Why can't we do `*` here?

File: ui/packages/catalog-ui-search/package.json [hunk 2]
```diff
@@ -25,6 +25,7 @@
     "node": ">=0.10.5"
   },
   "devDependencies": {
+    "@types/styled-components": "4.0.3",
```
Comment (Bdthomson) [inline ui/packages/catalog-ui-search/package.json]: Styled doesn't contain it's type definitions in the base dependency?

File: ui/packages/catalog-ui-search/package.json [hunk 3]
```diff
@@ -33,10 +34,10 @@
     "js-beautify": "1.8.7",
     "jsdom": "11.12.0",
     "loader-utils": "1.1.0",
-    "sinon": "5.0.7"
+    "sinon": "5.0.7",
+    "typescript": "3.5.3"
   },
   "dependencies": {
-    "@connexta/atlas": "0.0.35",
```
Comment (Bdthomson) [inline ui/packages/catalog-ui-search/package.json]: Why remove?

File: ui/packages/catalog-ui-search/src/main/webapp/react-component/container/loading-companion/loading-companion.tsx [hunk 4]
```diff
@@ -50,10 +50,6 @@ export default class LoadingCompanionContainer extends React.Component<
     LoadingCompanionView.stopLoadingElement(this.ref.current)
   }
   render() {
-    return (
-      <Root innerRef={this.ref as React.RefObject<HTMLDivElement>}>
-        {this.props.children}
-      </Root>
-    )
+    return <Root ref={this.ref as any}>{this.props.children}</Root>
```
Comment (Bdthomson) [inline ui/packages/catalog-ui-search/src/main/webapp/react-component/container/loading-companion/loading-companion.tsx]: Do we need a ticket made for retyping the refs after the styled typing issue is resolved?

File: ui/packages/catalog-ui-search/src/main/webapp/react-component/presentation/dropdown/dropdown.tsx [hunk 5]
```diff
@@ -100,10 +101,11 @@ export type withContext = {
   dropdownContext: Context
 }
 
-type Props = {
+export type Props = {
```
Comment (Bdthomson) [inline ui/packages/catalog-ui-search/src/main/webapp/react-component/presentation/dropdown/dropdown.tsx]: Should we rename to `DropdownProps` now that it's being exported?

File: ui/packages/catalog-ui-search/src/main/webapp/react-component/presentation/workspace-menu/workspace-menu.tsx [hunk 6]
```diff
@@ -44,7 +44,7 @@ const StyledWorkspaceTitle = styled.div`
   display: block;
 `
 
-const StyledDropdown = styled(Dropdown)`
+const StyledDropdown = styled(Dropdown as any)`
```
Comment (Bdthomson) [inline ui/packages/catalog-ui-search/src/main/webapp/react-component/presentation/workspace-menu/workspace-menu.tsx]: 😢

File: ui/packages/catalog-ui-search/typings/styled-components/index.d.tsx [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+ * License, or any later version.
+ *
+ * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without
+ * even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * Lesser General Public License for more details. A copy of the GNU Lesser General Public License
+ * is distributed along with this program and can be found at
+ * <http://www.gnu.org/licenses/lgpl.html>.
+ *
+ **/
+// import original module declarations
```
Comment (Bdthomson) [inline ui/packages/catalog-ui-search/typings/styled-components/index.d.tsx]: Remove?

File: ui/packages/catalog-ui-search/package.json (hunk 2 above)
Comment (andrewkfiedler) [reply ui/packages/catalog-ui-search/package.json]: No, a lot of libraries move their definitions to @typed.  They did with v4 as well.  You'll notice this isn't the latest version though: https://github.com/styled-components/styled-components-website/issues/447 

If we have bandwidth it'd be nice to help them fix up their types so we can use the latest type definitions they have.  But for now it'll mess up theming.

File: ui/packages/catalog-ui-search/package.json (hunk 3 above)
Comment (andrewkfiedler) [reply ui/packages/catalog-ui-search/package.json]: We no longer have any dependency on it in DDF.  @blen-desta removed the login usage.

File: ui/packages/catalog-ui-search/src/main/webapp/react-component/presentation/dropdown/dropdown.tsx (hunk 5 above)
Comment (andrewkfiedler) [reply ui/packages/catalog-ui-search/src/main/webapp/react-component/presentation/dropdown/dropdown.tsx]: Importers can rename it when they import it.  I might remove this though, I don't know if I ended up needing it to actually be exported.

File: ui/packages/catalog-ui-search/typings/styled-components/index.d.tsx (hunk 7 above)
Comment (andrewkfiedler) [reply ui/packages/catalog-ui-search/typings/styled-components/index.d.tsx]: They're helpful for future maintainers. 🤷‍♂

File: ui/packages/catalog-ui-search/src/main/webapp/react-component/container/loading-companion/loading-companion.tsx (hunk 4 above)
Comment (andrewkfiedler) [reply ui/packages/catalog-ui-search/src/main/webapp/react-component/container/loading-companion/loading-companion.tsx]: I think we'd have to help the styled type definitions out to actually get this fixed.

File: ui/package.json (hunk 1 above)
Comment (andrewkfiedler) [reply ui/package.json]: We're trying to force a specific version.  I'm not sure you can use * in a resolutions declaration.

**Comment by andrewkfiedler** [comment]
build now

**Comment by jordanwilking** [comment]
🍳 Hero Successful 🍳 

#### Steps

1. Did a full build and ran ddf
1. Ingested data
1. Confirmed notifications popped up and cleared them
1. Hid ingested data from future searches and confirmed it worked without style issues
1. Went into settings and cleared hidden items and confirmed that items were shown again  without style issues
- Generally kept an eye out for unexpected console errors and styling issues

#### Hero demo

DDFBuild
image

**Comment by andrewkfiedler** [comment]
Since this has been verified locally as building / working, I'm going to go around the build failure.  After talking with others, they say the build has been failing intermittently like this for weeks.

**Pull request merged**